File: 351001/Strizhak/rest_api/rest_api/Services/ReactionService.cs
using rest_api.Entities;
using rest_api.InMemory;
using rest_api.Dtos;

namespace rest_api.Services
{
    public class ReactionService : BaseService<Reaction, ReactionRequestTo, ReactionResponseTo>
    {
        public ReactionService(IRepository<Reaction> repository) : base(repository)
        {
        }

        public override ReactionResponseTo Create(ReactionRequestTo request)
        {
            var reaction = MapToEntity(request);
  

            _repository.Add(reaction);
            return MapToResponse(reaction);
        }

        public override ReactionResponseTo Update(ReactionRequestTo request)
        {
            var id = request.Id;
            var reaction = _repository.GetById(id);
            if (reaction == null)
                throw new KeyNotFoundException($"Reaction with id {id} not found");

            reaction.TopicId = request.TopicId;
            reaction.Content = request.Content;
            

            _repository.Update(reaction);
            return MapToResponse(reaction);
        }

        protected override ReactionResponseTo MapToResponse(Reaction entity)
        {
            return new ReactionResponseTo
            {
                Id = entity.Id,
                TopicId = entity.TopicId,
                Content = entity.Content,
                
            };
        }

        protected override Reaction MapToEntity(ReactionRequestTo request)
        {
            return new Reaction
            {
                TopicId = request.TopicId,
                Content = request.Content
            };
        }
    }
}

File: 351001/shaminko/module-discussion/src/main/java/by/shaminko/distcomp/services/ReactionService.java
package by.shaminko.distcomp.services;

import by.shaminko.distcomp.dto.ReactionMapper;
import by.shaminko.distcomp.dto.ReactionRequestTo;
import by.shaminko.distcomp.dto.ReactionResponseTo;
import by.shaminko.distcomp.dto.*;
import by.shaminko.distcomp.entities.Reaction;
import by.shaminko.distcomp.repositories.ReactionRepository;
import lombok.AllArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.UUID;

@Service
@AllArgsConstructor
public class ReactionService {
    private static final String APPROVED_STATE = "APPROVE";
    private static final String DECLINED_STATE = "DELCINE";
    private static final List<String> STOP_WORDS = List.of("spam", "scam", "fraud", "hate");

    public final ReactionRepository repImpl;
    @Qualifier("reactionMapper")
    public final ReactionMapper mapper;

    public List<ReactionResponseTo> getAll() {
        return repImpl.getAll().map(mapper::out).toList();
    }

    public ReactionResponseTo getById(Long id) {
        return repImpl.get(id).map(mapper::out)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found"));
    }

    public ReactionResponseTo create(ReactionRequestTo req) {
        Reaction reaction = mapper.in(req);
        reaction.setId(Math.abs(UUID.randomUUID().getMostSignificantBits()));
        reaction.setState(moderate(reaction.getContent()));
        return repImpl.create(reaction).map(mapper::out)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR));
    }

    public ReactionResponseTo update(ReactionRequestTo req) {
        if (repImpl.get(req.getId()).isEmpty()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Reaction not found");
        }
        Reaction reaction = mapper.in(req);
        reaction.setState(moderate(reaction.getContent()));
        return repImpl.update(reaction).map(mapper::out)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR));
    }

    public void delete(Long id) {
        repImpl.delete(id);
    }

    private String moderate(String content) {
        String normalized = content == null ? "" : content.toLowerCase();
        return STOP_WORDS.stream().anyMatch(normalized::contains) ? DECLINED_STATE : APPROVED_STATE;
    }
}